File: app/src/main/java/com/my/televip/features/ui/DisableNumberRounding.java
package com.my.televip.features.ui;

import com.my.televip.Class.ClassLoad;
import com.my.televip.Class.ClassNames;
import com.my.televip.Configs.ConfigManager;
import com.my.televip.base.BaseMethodHook;
import com.my.televip.hooks.HMethod;
import com.my.televip.logging.Logger;
import com.my.televip.obfuscate.ArgsResolver;
import com.my.televip.obfuscate.Obfuscate;

public class DisableNumberRounding {

    public static boolean isEnable = false;

    public static void init() {
        try {
            if (!isEnable) {
                isEnable = true;

                if (ClassLoad.getClass(ClassNames.LOCALE_CONTROLLER) != null) {
                    HMethod.hookMethod(ClassLoad.getClass(ClassNames.LOCALE_CONTROLLER), Obfuscate.getMethodName("LocaleController", "formatShortNumber"), ArgsResolver.merge("formatShortNumber", new Class[]{int.class, int[].class},
                            new BaseMethodHook() {
                                @Override
                                protected void beforeMethod(MethodHookParam param) {
                                    if (ConfigManager.disableNumberRounding.isEnable()) {
                                        int[] rounded = (int[]) param.args[1];
                                        int number = (int) param.args[0];
                                        if (rounded != null) {
                                            rounded[0] = number;
                                        }
                                        param.setResult(String.valueOf(number));
                                    }
                                }
                            }));
                }
            }
        } catch (Throwable t) {
            Logger.e(t);
        }
    }
}

File: app/src/main/java/com/my/televip/features/ui/DisableNumberRounding.kt
package com.my.televip.features.ui

import com.my.televip.clazz.ClassLoad
import com.my.televip.clazz.ClassNames
import com.my.televip.configs.ConfigManager
import com.my.televip.base.BaseMethodHook
import com.my.televip.logging.Logger
import com.my.televip.obfuscate.ArgsResolver
import com.my.televip.obfuscate.Obfuscate
import de.robv.android.xposed.XposedHelpers

object DisableNumberRounding {

    @JvmField
    var isEnable: Boolean = false

    @JvmStatic
    fun init() {
        try {
            if (!isEnable) {
                isEnable = true
                val lcClass = ClassLoad.getClass(ClassNames.LOCALE_CONTROLLER)
                if (lcClass != null) {
                    val types: Array<Class<*>> = arrayOf(
                        Int::class.javaPrimitiveType!!,
                        IntArray::class.java
                    )
                    val merged = ArgsResolver.merge("formatShortNumber", types, object : BaseMethodHook() {
                        override fun beforeMethod(param: MethodHookParam) {
                            if (ConfigManager.disableNumberRounding?.isEnable == true) {
                                val rounded = param.args[1] as? IntArray
                                val number = param.args[0] as Int
                                if (rounded != null && rounded.isNotEmpty()) {
                                    rounded[0] = number
                                }
                                param.result = number.toString()
                            }
                        }
                    })
                    if (merged != null) {
                        XposedHelpers.findAndHookMethod(
                            lcClass,
                            Obfuscate.getMethodName("LocaleController", "formatShortNumber"),
                            *merged
                        )
                    }
                }
            }
        } catch (t: Throwable) {
            Logger.e(t)
        }
    }
}
